stdlib/array.ml

```diff
@@ -292,3 +292,27 @@ let stable_sort cmp a =
 
 
 let fast_sort = stable_sort
+
+(** {6 Iterators} *)
+
+let to_iter a =
+  let next i =
+    if i < length a
+    then
+      let x = unsafe_get a i in
+      Iter.Yield (x, i+1)
+    else Iter.Done
+  in Iter.Sequence (0, next)
+
+let to_iteri a =
+  let next i =
+    if i < length a
+    then
+      let x = unsafe_get a i in
+      Iter.Yield ((i,x), i+1)
+    else Iter.Done
+  in Iter.Sequence (0, next)
+
+let of_iter i =
+  let l = Iter.fold_left (fun acc x -> x::acc) [] i in
+  of_list (List.rev l)
```

Comment:
There is a comment about `List` depending on `Array` to justify a local definition of `list_length` in array.ml.  This does not seem to be true (any more, see the commented implementation of `stable_sort` in list.ml).  Anyway, one should perhaps avoiding introducing a dependency from `List` to `Array` in case we need back the other dependency.

But the best is perhaps to add a of_list_rev to avoid the extra list copying.

Another implementation could also completely get rid of any intermediate heap allocations by using stack storage (up to a fixed depth).